### PR TITLE
[EMB-205] Decorate the models

### PR DESCRIPTION
**Pull request by aaxelb**
## Purpose
Use decorators (and ES6 classes) for all our models.

## Summary of Changes
- Decorators!
- Rename `contributor.users` to `user`, because there's only one.
- Refactor `osf-model.queryHasMany` to use a task.

## Side Effects / Testing Notes
Everything should be the same.

## Ticket

https://openscience.atlassian.net/browse/EMB-205

# Reviewer Checklist

- [ ] meets requirements
- [ ] easy to understand
- [ ] DRY
- [ ] testable and includes test(s)
- [ ] changes described in `CHANGELOG.md`

### CODE REVIEW

File: app/models/citation.ts [hunk 1]
```diff
@@ -13,9 +11,9 @@ const { attr } = DS;
  *
  * @class Citation
  */
-export default class Citation extends OsfModel.extend({
-    citation: attr('string'),
-}) {}
+export default class Citation extends OsfModel {
+    @attr('string') citation;
```
Comment (jamescdavis) [inline app/models/citation.ts]: We should also probably type attributes:
```
@attr('string') citation: Computed<string>;
```

File: app/models/contributor.ts [hunk 2]
```diff
@@ -15,48 +12,19 @@ const { attr, belongsTo } = DS;
...
-            if (this.get('isNew')) {
-                return this.get('_userId');
-            } else {
-                return this.get('id').split('-').pop();
-            }
-        },
-        set(_: any, userId: string): void {
-            this.set('_userId', userId);
-        },
-    }).volatile();
```
Comment (jamescdavis) [inline app/models/contributor.ts]: It looks like the serializer [uses this](https://github.com/CenterForOpenScience/ember-osf-web/blob/develop/app/serializers/contributor.ts#L24).

File: app/models/contributor.ts [hunk 3]
```diff
@@ -15,48 +12,19 @@ const { attr, belongsTo } = DS;
...
-            if (this.get('isNew')) {
-                return this.get('_nodeId');
-            } else {
-                return this.get('id').split('-').shift();
-            }
-        },
-        set(_: any, nodeId: string): void {
-            this.set('_nodeId', nodeId);
-        },
-    }).volatile();
```
Comment (jamescdavis) [inline app/models/contributor.ts]: ...and the adapter [uses this](https://github.com/CenterForOpenScience/ember-osf-web/blob/develop/app/adapters/contributor.ts#L16).

**Comment by jamescdavis** [comment]
There are a number of [TypeScript errors](https://travis-ci.org/CenterForOpenScience/ember-osf-web/builds/366292329#L712-L735) in the models. A few are due to mixins, but I think some of these can be resolved?

File: app/models/collection.ts [hunk 4]
```diff
@@ -1,34 +1,36 @@
...
+    @attr('date') dateModified: Date;
+    @attr('boolean') bookmarks: boolean;
+
+    @hasMany('node', {
         inverse: null,
         serializerType: 'linked-node',
-    }),
-    linkedRegistrations: hasMany('registration', {
+    })
+    linkedNodes: Node[];
```
Comment (jamescdavis) [inline app/models/collection.ts]: Should probably type this: `DS.PromiseManyArray<Node>`

Comment (jamescdavis) [reply app/models/collection.ts]: (and all the other hasManys)

File: app/models/preprint-provider.ts [hunk 5]
```diff
@@ -1,38 +1,42 @@
...
-    highlightedTaxonomies: hasMany('taxonomy'),
-    preprints: hasMany('preprint', { inverse: 'provider', async: true }),
-    licensesAcceptable: hasMany('license', { inverse: null }),
-}) {
-    reviewableStatusCounts = alias('links.relationships.preprints.links.related.meta');
-    // tslint:disable-next-line max-line-length
-    hasHighlightedSubjects = alias('links.relationships.highlighted_taxonomies.links.related.meta.has_highlighted_subjects'); // eslint-disable-line max-len
+    @hasMany('taxonomy') taxonomies: Taxonomy[];
+    @hasMany('taxonomy') highlightedTaxonomies: Taxonomy[];
+    @hasMany('preprint', { inverse: 'provider', async: true }) preprints: Preprint[];
```
Comment (jamescdavis) [inline app/models/preprint-provider.ts]: Since `async: true` is default, should we just remove all these?

**Comment by icereval** [comment]
+1

File: app/adapters/contributor.ts [hunk 6]
```diff
@@ -1,43 +1,7 @@
...
-            return `${base}${id.split('-').pop()}/`;
-        }
-
-        const sendEmail: boolean | string = snapshot ? snapshot.record.get('sendEmail') : true;
-
-        // Needed for Ember Data to update the inverse record's (the node's) relationship
-        return `${base}?embed=node&send_email=${sendEmail}`;
-    },
-}) {}
+export default class Contributor extends OsfAdapter {
```
Comment (binoculars) [inline app/adapters/contributor.ts]: Cruft removal? Much cleaner looking. Not needed anymore?

File: types/ember-cp-validations/index.d.ts [hunk 7]
```diff
@@ -1,4 +1,20 @@
+interface ValidateOptions {
```
Comment (binoculars) [inline types/ember-cp-validations/index.d.ts]: 🔥

File: app/adapters/contributor.ts (hunk 6 above)
Comment (aaxelb) [reply app/adapters/contributor.ts]: We don't yet have anything that adds/removes/modifies contributors, so this stuff is unneeded. When we do add contributor management, we should go about it [a different way](https://github.com/CenterForOpenScience/ember-osf-web/pull/197).

**Comment by jamescdavis** [comment]
:fire: :metal: :star2:

**Pull request merged**